core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java

```diff
@@ -0,0 +1,393 @@
+                            )
+                    );
+                }
+            }
+
+            responses = new AtomicReferenceArray<>(nodeIds.size());
+        }
+
+        public void start() {
+            if (nodeIds.size() == 0) {
+                try {
+                    listener.onResponse(newResponse(request, new AtomicReferenceArray(0), unavailableShardExceptions, nodeIds));
+                } catch (Throwable e) {
+                    listener.onFailure(e);
+                }
+            } else {
+                int nodeIndex = -1;
+                for (Map.Entry<String, List<ShardRouting>> entry : nodeIds.entrySet()) {
+                    nodeIndex++;
+                    DiscoveryNode node = nodes.get(entry.getKey());
+                    sendNodeRequest(node, entry.getValue(), nodeIndex);
+                }
+            }
+        }
+
+        private void sendNodeRequest(final DiscoveryNode node, List<ShardRouting> shards, final int nodeIndex) {
+            try {
+                BroadcastByNodeRequest nodeRequest = newNodeRequest(node.getId(), request, shards);
+                transportService.sendRequest(node, transportNodeBroadcastAction, nodeRequest, new BaseTransportResponseHandler<BroadcastByNodeResponse>() {
+                    @Override
+                    public BroadcastByNodeResponse newInstance() {
+                        return newNodeResponse();
+                    }
+
+                    @Override
+                    public void handleResponse(BroadcastByNodeResponse response) {
+                        onNodeResponse(node, nodeIndex, response);
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        onNodeFailure(node, nodeIndex, exp);
+                    }
+
+                    @Override
+                    public String executor() {
+                        return ThreadPool.Names.SAME;
+                    }
+                });
+            } catch (Throwable e) {
+                onNodeFailure(node, nodeIndex, e);
+            }
+        }
+
+        protected void onNodeResponse(DiscoveryNode node, int nodeIndex, BroadcastByNodeResponse response) {
+            logger.trace("received response from node [{}]", node.id());
+            responses.set(nodeIndex, response);
+            if (counter.incrementAndGet() == responses.length()) {
+                onCompletion();
+            }
+        }
+
+        protected void onNodeFailure(DiscoveryNode node, int nodeIndex, Throwable t) {
+            String nodeId = node.id();
+            if (logger.isDebugEnabled() && !(t instanceof NodeShouldNotConnectException)) {
+                logger.debug("failed to execute on node [{}]: [{}]", nodeId, t);
+            }
+
+            responses.set(nodeIndex, new FailedNodeException(nodeId, "Failed node [" + nodeId + "]", t));
```

Comment:
can we use compareAndSet here , make sure it's null and only then increment the response counter? This way we're protected from double handling of failure / responses (needed because of the throwable handling in L308)